### DISCUSSION TITLE
References

### DISCUSSION THREAD

**Discussion by jmafc** [hsutter/cppfront#593] [Q&A] [answered]
Is there a way of declaring plain reference variables in CPP2? e.g.,
```c++
int foo = 10;
int& bar = foo;
```
I tried `bar: i32& = foo;` but got `error: missing semicolon at end of declaration or equal at start of initializer (at '&')`

**Comment by AbhinavK00** [comment, marked as answer]
Cppfront intends to do away with references completely. The basic guideline is to use pointers instead. If you do want refereces, you will have to use cpp style syntax.

**Reply by jmafc** [reply]
Ok, but if I say `bar: *i32 = foo&;`, the generated cpp is a raw pointer, which is very counter-intuitive, since we've been told countless times "raw pointers are unsafe, use smart pointers instead", yet cppfront says `all Cpp2, passes safety checks`.

I see also that if I try to change `bar`, e.g., `bar* = 42;` (I have to remember it's a pointer and dereference it, cppfront does catch that) and the generated code is ` *cpp2::assert_not_null(bar) = 42;` which presumably will give an exception and terminate if `bar` is somehow disconnected from `foo`. Is that the safety guarantee?

**Reply by AbhinavK00** [reply]
> since we've been told countless times "raw pointers are unsafe, use smart pointers instead",

That's not fully correct. Use smart pointers when you have to manage ownership, non-owning pointers are fine (as long as they don't outlive the owner).

**Comment by JohelEGP** [comment]
You can use a parameterized statement to the same effect (<https://cpp2.godbolt.org/z/dqPaTP55W>):
```Cpp2
main: () = {
  i := 0;
  (inout j := i) {
    [[assert: i& == j&]]
  }
}
```

**Reply by jmafc** [reply]
Is the "Pascal/Go"ish `:=` the equivalent of `auto`?  Is this described somewhere other than examples? (Admittedly, I haven't watched all of the talks or looked at all the regression tests).

**Reply by JohelEGP** [reply]
See <https://github.com/hsutter/cppfront/issues/386#issuecomment-1518194517>.

**Reply by JohelEGP** [reply]
`x := 0;` is a shorthand for
`x: _ = 0;`,
where `_` is the unconstrained placeholder "`auto`", if that makes sense.
